src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -468,4 +501,195 @@ UA_PubSubKeyStorage_removeKeyStorage(UA_Server *server, UA_PubSubKeyStorage *key
+        if(retval != UA_STATUSCODE_GOOD)
+            goto cleanup;
+    }
+
+    UA_Duration msTimeToNextKey =
+        *(UA_Duration *)response->results->outputArguments[3].data;
+    if(!(msTimeToNextKey > 0))
+        msTimeToNextKey = ks->keyLifeTime;
+    retval = UA_PubSubKeyStorage_addKeyRolloverCallback(
+        server, ks, (UA_ServerCallback)UA_PubSubKeyStorage_keyRolloverCallback,
+        msTimeToNextKey, &ks->callBackId);
+
+cleanup:
+    *ctx->loading = false;
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                     "Failed to store the fetched keys from SKS server with error: %s",
+                     UA_StatusCode_name(retval));
+    }
+    if(ctx)
+        UA_free(ctx);
+    return;
+}
+
+static UA_StatusCode
+callGetSecurityKeysMethod(UA_Server *server, UA_Client *sksClient,
+                          UA_PubSubKeyStorage *ks, UA_UInt32 startingTokenId,
+                          UA_UInt32 requestedKeyCount, UA_Boolean *loadingKeys) {
+    UA_Variant *inputArguments = (UA_Variant *)UA_calloc(3, (sizeof(UA_Variant)));
+    UA_Variant_setScalarCopy(&inputArguments[0], &ks->securityGroupID,
+                             &UA_TYPES[UA_TYPES_STRING]);
+    UA_Variant_setScalarCopy(&inputArguments[1], &startingTokenId,
+                             &UA_TYPES[UA_TYPES_UINT32]);
+    UA_Variant_setScalarCopy(&inputArguments[2], &requestedKeyCount,
+                             &UA_TYPES[UA_TYPES_UINT32]);
+
+    UA_NodeId objectId = UA_NODEID_NUMERIC(0, UA_NS0ID_PUBLISHSUBSCRIBE);
+    UA_NodeId methodId = UA_NODEID_NUMERIC(0, UA_NS0ID_PUBLISHSUBSCRIBE_GETSECURITYKEYS);
+    size_t inputArgumentsSize = 3;
+
+    getSecurityKeysContext *ctx =
+        (getSecurityKeysContext *)UA_malloc(sizeof(getSecurityKeysContext));
+    ctx->ks = ks;
+    ctx->server = server;
+    ctx->loading = loadingKeys;
+    return UA_Client_call_async(sksClient, objectId, methodId, inputArgumentsSize,
+                                inputArguments, storeFetchedKeys, (void *)ctx, NULL);
+}
+
+static void
+onConnect(UA_Client *client, UA_SecureChannelState channelState,
+          UA_SessionState sessionState, UA_StatusCode connectStatus) {
+    if(connectStatus != UA_STATUSCODE_GOOD && channelState != UA_SECURECHANNELSTATE_CLOSED) {
+        UA_LOG_ERROR(&client->config.logger, UA_LOGCATEGORY_CLIENT,
+                    "Failed to connect SKS server with error: %s ",
+                    UA_StatusCode_name(connectStatus));
+        client->noReconnect = true;
+    }
+}
+
+UA_StatusCode
+getSecurityKeysAndStoreFetchedKeys(UA_Server *server, UA_PubSubKeyStorage *keyStorage) {
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+    UA_UInt32 startingTokenId = UA_REQ_CURRENT_TOKEN;
+    UA_UInt32 requestKeyCount = UA_UINT32_MAX;
+
+    UA_ClientConfig *cc = keyStorage->sksConfig.sksClientConfig;
+    cc->stateCallback = onConnect;
+    UA_Client *client = UA_Client_newWithConfig(cc);
```

Comment:
> This moves the config into the client. As such the config you give this will be cleaned up automatically. That should be documented.
> 

ACK 
> Or we need some UA_ClientConfig_copy. There already exists UA_ClientConfig_clear which can also be exposed as public API together with _copy.

I was also thinking along the same lines. However, I would like to keep the scope of the PR clear. I will create a separate PR for these APIs and use them  here.